### PR TITLE
Overriding default and setting up GWC with jdbc for docker.

### PR DESCRIPTION
**Pull request by RemiDesgrange**
For docker setup. JDBC is now configured for geoserver. 

The default value is still H2 when geoserver first boot, but you can change it in the interface.
image

### CODE REVIEW

File: geoserver/webapp/src/docker/jetty-env.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+<Configure class="org.eclipse.jetty.webapp.WebAppContext">
+   <New id="ds_gwc_geoserver" class="org.eclipse.jetty.plus.jndi.Resource">
+     <Arg>jdbc/gwc</Arg>
+     <Arg>
+       <New class="org.apache.commons.dbcp.BasicDataSource">
+         <Set name="driverClassName">org.postgresql.Driver</Set>
+         <Set name="username">georchestra</Set>
+         <Set name="password">georchestra</Set>
+         <Set name="url">jdbc:postgresql://database/georchestra</Set>
+         <Set name="validationQuery">SELECT 1</Set>
```
Comment (pmauduit) [inline geoserver/webapp/src/docker/jetty-env.xml:12]: I don't know if this is relevant here, but we could also use env variables (too bad that we should also have the info in the default.properties from the georchestra datadir)

File: docs/setup/postgresql.md [hunk 2]
```diff
@@ -54,6 +54,11 @@ If the **Atlas** is deployed:
 psql -d georchestra -f postgresql/060-atlas.sql
 ```
 
+## GeoWebCache schema
+
+If geowebcache is used it creates a geowebcache schema to store quota and relevant infos to
```
Comment (pmauduit) [inline docs/setup/postgresql.md:59]: "is used, ..." (adding a coma)

File: docs/setup/postgresql.md [hunk 3]
```diff
@@ -54,6 +54,11 @@ If the **Atlas** is deployed:
...
+geowebcache. This setup is highly encouraged in a production environment to replace local H2 database.
```
Comment (pmauduit) [inline docs/setup/postgresql.md:60]: "... to replace the local H2 database"

File: geoserver/webapp/src/docker/Dockerfile [hunk 4]
```diff
@@ -28,11 +28,17 @@ RUN mkdir /mnt/geoserver_datadir /mnt/geoserver_geodata /mnt/geoserver_tiles /mn
 
 # add a tweaked configuration. First use case is support of OPTIONS.
 ADD web.xml /var/lib/jetty/webapps/geoserver/WEB-INF/web.xml
+ADD jetty-env.xml /var/lib/jetty/webapps/geoserver/WEB-INF/jetty-env.xml
+# tweaked geowebcache config that use postgresql instead of H2. Copying it to /tmp since copying it
```
Comment (pmauduit) [inline geoserver/webapp/src/docker/Dockerfile:32]: " ... which uses postgresql"

File: geoserver/webapp/src/docker/Dockerfile [hunk 5]
```diff
@@ -28,11 +28,17 @@ RUN mkdir /mnt/geoserver_datadir /mnt/geoserver_geodata /mnt/geoserver_tiles /mn
 
 # add a tweaked configuration. First use case is support of OPTIONS.
 ADD web.xml /var/lib/jetty/webapps/geoserver/WEB-INF/web.xml
+ADD jetty-env.xml /var/lib/jetty/webapps/geoserver/WEB-INF/jetty-env.xml
+# tweaked geowebcache config that use postgresql instead of H2. Copying it to /tmp since copying it 
+# to /mnt/geoserver_tiles won't work (it's a volume, data will be override). File copied to
```
Comment (pmauduit) [inline geoserver/webapp/src/docker/Dockerfile:33]: "data will be overridden). Instead, the file will be copied to the actual destination by the entrypoint script.

File: geoserver/webapp/src/docker/docker-entrypoint.d/00-geoserver-bootstrap [hunk 6]
```diff
@@ -16,4 +16,9 @@ else
 
     echo 'Change log config according to docker setup'
     sed -i 's:stdOutLogging>false:stdOutLogging>true:g' /mnt/geoserver_datadir/logging.xml
+
+    # geowebcache config need to be placed in geoserver_tiles volume.
+    # copying it directly in the Dockerfile doesn't work. Copying it at runtime.
```
Comment (pmauduit) [inline geoserver/webapp/src/docker/docker-entrypoint.d/00-geoserver-bootstrap:21]: geowebcache config needs to be placed into the geoserver_tiles volume.
We cannot copy the file via the Dockerfile at build time, so we are doing it at runtime.

File: geoserver/webapp/src/docker/jetty-env.xml (hunk 1 above)
Comment (RemiDesgrange) [reply geoserver/webapp/src/docker/jetty-env.xml:12]: Your right. I'll update to env var + default values.

**Comment by fvanderbiest** [comment]
@RemiDesgrange up for the release ? Thanks :-)

**Comment by fvanderbiest** [comment]
> @RemiDesgrange up for the release ? Thanks :-)

Oh, forget it, I'm not releasing master ... yet ;-)

File: geoserver/webapp/src/docker/Dockerfile [hunk 7]
```diff
@@ -28,11 +28,17 @@ RUN mkdir /mnt/geoserver_datadir /mnt/geoserver_geodata /mnt/geoserver_tiles /mn
...
+# /mnt/geoserver_tiles by a entrypoint script.
```
Comment (fvanderbiest) [inline geoserver/webapp/src/docker/Dockerfile:34]: If the image provides `/mnt/geoserver_tiles/geowebcache-diskquota.xml` and `/mnt/geoserver_tiles/geowebcache-diskquota-jdbc.xml`, the first time an empty volume is mounted on it, it shall be initialized with the content provided by the image, which is exactly what we want to achieve here IIUC ?

Comment (fvanderbiest) [reply geoserver/webapp/src/docker/Dockerfile:34]: No need to use /tmp then

File: docs/setup/postgresql.md (hunk 3 above)
Comment (fvanderbiest) [reply docs/setup/postgresql.md:60]: Missing psql statement with 110-geowebcache.sql  (for those who do the install manually)

File: geoserver/webapp/src/docker/geowebcache-diskquota-jdbc.xml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<gwcJdbcConfiguration>
+  <dialect>PostgreSQL</dialect>
+  <JNDISource>jdbc/gwc</JNDISource>
+  <connectionPool>
+    <driver>org.postgresql.Driver</driver>
+    <url>jdbc:postgresql://database/georchestra</url>
+    <username>georchestra</username>
+    <password>georchestra</password>
+    <minConnections>1</minConnections>
```
Comment (fvanderbiest) [inline geoserver/webapp/src/docker/geowebcache-diskquota-jdbc.xml:9]: How does GWC know it shall use the `geowebcache` schema to store its tables ?

**Review by fvanderbiest — CHANGES_REQUESTED** [review]
Some gotchas ...

**Comment by arnaud-morvan** [comment]
I moved the config in geoserver_minimal_datadir: https://github.com/georchestra/geoserver_minimal_datadir/pull/14

**Pull request merged**